scio-jdbc/src/main/scala/com/spotify/scio/jdbc/sharded/JdbcShardable.scala

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright 2020 Spotify AB.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.scio.jdbc.sharded
+
+import java.sql.ResultSet
+import org.slf4j.LoggerFactory
+
+trait JdbcShardable[T] extends Serializable {
+
+  def columnValueDecoder(resultSet: ResultSet, columnName: String): T
+
+  def partition(range: Range[T], numShards: Int): Seq[ShardQuery]
+
+}
+
+object ShardBy {
+
+  object range {
+    def of[T: RangeJdbcShardable]: JdbcShardable[T] = implicitly[RangeJdbcShardable[T]]
+  }
+
+  object prefix {
+    def of[T](
+      prefixLength: Int
+    )(implicit shardeableF: Int => PrefixJdbcShardable[T]): JdbcShardable[T] =
+      shardeableF(prefixLength)
+  }
+
+}
+
+object JdbcShardable {
+
+  implicit val longRangeJdbcShardable: RangeJdbcShardable[Long] = RangeJdbcShardable[Long](
+    (rs, colName) => rs.getLong(colName),
+    (range, numShards) => (range.upperBound - range.lowerBound) / numShards
+  )
+
+  implicit val intRangeJdbcShardable: RangeJdbcShardable[Int] = RangeJdbcShardable[Int](
+    (rs, colName) => rs.getInt(colName),
+    (range, numShards) => (range.upperBound - range.lowerBound) / numShards
+  )
+
+  implicit val decimalRangeJdbcShardable: RangeJdbcShardable[BigDecimal] =
+    RangeJdbcShardable[BigDecimal](
+      (rs, colName) => rs.getBigDecimal(colName),
+      (range, numShards) => (range.upperBound - range.lowerBound) / numShards
+    )
+
+  implicit val doubleRangeJdbcShardable: RangeJdbcShardable[Double] =
+    RangeJdbcShardable[Double](
+      (rs, colName) => rs.getDouble(colName),
+      (range, numShards) => (range.upperBound - range.lowerBound) / numShards
+    )
+
+  implicit val floatRangeJdbcShardable: RangeJdbcShardable[Float] =
+    RangeJdbcShardable[Float](
+      (rs, colName) => rs.getFloat(colName),
+      (range, numShards) => (range.upperBound - range.lowerBound) / numShards
+    )
+
+  implicit val stringPrefixJdbcShardable: Int => PrefixJdbcShardable[String] = prefixLength =>
```

Comment:
Since this is still not implemented, I guess it's better if we don't supply the implicit? or will you still implement it in this PR?

Reply:
I think sharding by an arbitrary range of strings deserves its own PR and might take some time to implement it properly, so probably would remove it from this PR.